### ISSUE TITLE
Keywords in search and filtering (EUTF) (estimate: 16)

### ISSUE THREAD

**Issue by nadiagorchakova** [akvo/akvo-rsr#2548]
**Context**

EUTF programme managers need a better overview of regions and delegations they are in charge of. 

**Problem**

The available advanced filtering feature does not cover their organisation-specific needs. For instance, location filter does not have the location categories they use at the EU, for instance, Sahel region.
It makes work with RSR difficult for them as being in charge of tens of projects, it's impossible to get a high-level overview without a meaningful filter/search solution.

**Suggested solution** 

Use keywords to cater for their organisation-specific filtering needs. Keyword option should be added to the advanced filter and into the search field algorithm.

In their Akvo app page they should only see their own keywords so that it's easier for the final users to pick the right one from the list.

As a more advanced option (probably for later), filter should be dynamic in a way that if you select just one implementing partner as a keyword, the options for location should be narrowed down to the ones this implementing partner is active in.

**Comment by nadiagorchakova** [comment]
I provisionally put this issue into the next iteration but it's up for discussion.

**Comment by nadiagorchakova** [comment]
This part should be done as part of a separate issue:

"As a more advanced option (probably for later), filter should be dynamic in a way that if you select just one implementing partner as a keyword, the options for location should be narrowed down to the ones this implementing partner is active in."

**Comment by zzgvh** [comment]
Hey @punchagan I started looking at the typeahead and fell into a rabbit hole with a bear trap! Meaning I've written some code to support the typeahead work, but I got stuck on the first simple test case for the new PartnerSite methods failing and I don't understand what's going on. I'll try to describe why I ended up writing this code...

The typeaheads on the page are driven by these calls: [orgs](https://github.com/akvo/akvo-rsr/blob/4b293f036304728b919bbc968a99b46bdceacb3c/akvo/rsr/static/scripts-src/project-directory.js#L40), [projects](https://github.com/akvo/akvo-rsr/blob/4b293f036304728b919bbc968a99b46bdceacb3c/akvo/rsr/static/scripts-src/project-directory.js#L67), [locations](https://github.com/akvo/akvo-rsr/blob/4b293f036304728b919bbc968a99b46bdceacb3c/akvo/rsr/static/scripts-src/project-directory.js#L52). (Note that there's another typeahead in the advanced filter box for organisations too.) But the ones above drive the typeahead for the search field above the list of projects. The problem is that only the project typeahead API call filters the result. ```/rest/v1/typeaheads/organisations```returns all orgs!

So I thought I'd fix the org typeahead API call to filter if we're on a Page. Hence [this code](https://github.com/akvo/akvo-rsr/blob/4b293f036304728b919bbc968a99b46bdceacb3c/akvo/rsr/models/partner_site.py#L170-L199). I added partners() and updates() for symmetry and with the intent to cleanup the [organisation](https://github.com/akvo/akvo-rsr/blob/4b293f036304728b919bbc968a99b46bdceacb3c/akvo/rsr/views/organisation.py#L38) and [updates](https://github.com/akvo/akvo-rsr/blob/4b293f036304728b919bbc968a99b46bdceacb3c/akvo/rsr/views/project_update.py#L48) listings.

Which brings me to where I'm stuck.

[The test](https://github.com/akvo/akvo-rsr/blob/4b293f036304728b919bbc968a99b46bdceacb3c/akvo/rsr/tests/models/test_partner_site.py) is to check that PartnerSite.projects() returns the correct number of projects, i.e. the projects associated with the PartnerSite. But things go wrong [here](https://github.com/akvo/akvo-rsr/blob/4b293f036304728b919bbc968a99b46bdceacb3c/akvo/rsr/tests/models/test_partner_site.py#L84). Here's some terminal output when I investigate at the breakpoint:

```-> projects_1 = page_1.projects()
(Pdb) page_1.projects()
[]
(Pdb) page_1.organisation.projects.all()
[<Project: Test project 2>, <Project: Test project 1>]
(Pdb) page_1.organisation.projects.all()[0].publishingstatus.status
u'unpublished'
(Pdb) self.project_2
<Project: Test project 2>
(Pdb) self.project_2.publishingstatus.status
'published'
```

Now I know I don't know much about how to write test, but what the !#%& is going on?!? Is there some transaction weirdness that prevents the save() on lines 31 and 34 to take? Or am I just too tired... 😫

Anyway, maybe you can continue. Or just rip it out and do something that works! :wink:

**Comment by punchagan** [comment]
> Now I know I don't know much about how to write test, but what the !#%& is going on?!? Is there some transaction weirdness that prevents the save() on lines 31 and 34 to take? Or am I just too tired...

Looks like you were just tired.  The `save` had to be called on the publishing status, not the project. :wink:

**Comment by punchagan** [comment]
Closed with #2591 

Feel free to re-open if changes need to be made to this.
